### ISSUE TITLE
Factory file fails to compile when type `uint256[]` is used in the constructor

### ISSUE THREAD

**Issue by naddison36** [ethereum-ts/TypeChain#168]
The generated `ConstructorWithIntArrayFactory.ts` file for the following Solidity will fail to compile as the `BigNumberish` type is not imported from `ethers/utils`

```Solidity
contract ConstructorWithIntArray {

    uint256[] someArray;

    constructor(uint256[] memory _someArray) public {
        someArray = _someArray;
    }
}
```

The generated `ConstructorWithIntArrayFactory.ts` file is
```ts
/* Generated by ts-generator ver. 0.0.8 */
/* tslint:disable */

import { Contract, ContractFactory, Signer } from "ethers";
import { Provider } from "ethers/providers";
import { UnsignedTransaction } from "ethers/utils/transaction";

import { ConstructorWithIntArray } from "./ConstructorWithIntArray";

export class ConstructorWithIntArrayFactory extends ContractFactory {
  constructor(signer?: Signer) {
    super(_abi, _bytecode, signer);
  }
  deploy(_someArray: (BigNumberish)[]): Promise<ConstructorWithIntArray> {
    return super.deploy(_someArray) as Promise<ConstructorWithIntArray>;
  }
  getDeployTransaction(_someArray: (BigNumberish)[]): UnsignedTransaction {
    return super.getDeployTransaction(_someArray);
  }
  attach(address: string): ConstructorWithIntArray {
    return super.attach(address) as ConstructorWithIntArray;
  }
  connect(signer: Signer): ConstructorWithIntArrayFactory {
    return super.connect(signer) as ConstructorWithIntArrayFactory;
  }
  static connect(
    address: string,
    signerOrProvider: Signer | Provider
  ): ConstructorWithIntArray {
    return new Contract(
      address,
      _abi,
      signerOrProvider
    ) as ConstructorWithIntArray;
  }
}

const _abi = [
  {
    inputs: [
      {
        internalType: "uint256[]",
        name: "_someArray",
        type: "uint256[]"
      }
    ],
    payable: false,
    stateMutability: "nonpayable",
    type: "constructor"
  }
];

const _bytecode =
  "0x608060405234801561001057600080fd5b5060405161018f38038061018f8339818101604052602081101561003357600080fd5b810190808051604051939291908464010000000082111561005357600080fd5b90830190602082018581111561006857600080fd5b825186602082028301116401000000008211171561008557600080fd5b82525081516020918201928201910280838360005b838110156100b257818101518382015260200161009a565b5050505091909101604052505082516100d492506000915060208401906100db565b5050610143565b828054828255906000526020600020908101928215610116579160200282015b828111156101165782518255916020019190600101906100fb565b50610122929150610126565b5090565b61014091905b80821115610122576000815560010161012c565b90565b603e806101516000396000f3fe6080604052600080fdfea265627a7a72315820311701576656af678908509fac0c466c381c9bd65f0f04fa674cdf92fb1ab24c64736f6c634300050b0032";
```

The typescript errors are
```
Error:(14, 23) TS2304: Cannot find name 'BigNumberish'.
Error:(17, 37) TS2304: Cannot find name 'BigNumberish'.
```

The workaround is to include a dummy constructor parameter of type `uint256` so the `BigNumberish` type is imported in the factory type definition file.

**Comment by naddison36** [comment]
The problem is with this line in the Ethers generator
https://github.com/ethereum-ts/TypeChain/blob/master/lib/targets/ethers/generation.ts#L84
```ts
if (constructorArgs.match(/: BigNumberish/)) ethersUtilsImports.push("BigNumberish");
```

For the above example, the `constructorArgs` will have value `_someArray: (BigNumberish)[], ` which will not match `/: BigNumberish/`. I'd suggest just matching on `/BigNumberish/` as the BigNumberish could be in a multidimensional array or a struct.

I was going to raise a PR but I can't get the tests to work before I do my change.

**Comment by krzkaczor** [comment]
Thanks, the proposed solution makes sense. @quezak anything to add? I think you were the one who wrote it in the first place?

@naddison36 i need to improve docs for running it locally but tests always pass on CI. You can take a look at our CI scripts here: https://github.com/ethereum-ts/TypeChain/blob/master/.circleci/config.yml

**Comment by quezak** [comment]
You're right, I didn't check that case before. It will probably fail for tuple types too.

@naddison36 since this is a simple change, you can try to submit a PR blindly hoping the tests will pass :joy: otherwise I'll do that tomorrow.

Matching only `/BigNumberish/` looks a bit too weak -- it will generate an unused import if your arg names include `BigNumberish`, but that's a really edge case (why would you include that in the arg name if its type is different ;) ).  But right now I don't see alternatives which use a regex that's reasonably easy to understand, so that's OK.

Please also note that the same fix seems to be needed for `Arrayish`.

**Comment by quezak** [comment]
Actually, this should be a bit better, -- not too complicated: `/\WBigNumberish(\W|$)/`
It should exclude all _sane_ false positives where `BigNumberish` is found in the arg name, but not in the types -- see last example here: https://regex101.com/r/WeFMHp/2

**Comment by naddison36** [comment]
@quezak good idea on checking for whitespace on either side of Arrayish and BigNumberish
I've raised PR https://github.com/ethereum-ts/TypeChain/pull/169

**Comment by krzkaczor** [comment]
Thanks for your PR @naddison36!

Sorry that I forgot to publish it right away. I just did it.

Published as `typechain@0.3.21`

**Comment by naddison36** [comment]
thanks @krzkaczor
